### PR TITLE
[SYSTEMML-831][WIP] Implement t-SNE algorithm

### PR DESCRIPTION
**Pull request by iyounus**
This is the first commit for tSNE in dml. At the moment this is using batch gradient method which will no scale. Next step is to implement stochastic gradient method to improve performance.

Its was tested in stand alone mode. The results from the algorithm are correct. I still need to test this on distributed data. No performance stats are available at this time.

@niketanpansare @dusenberrymw @mboehm7 @nakul02

### CODE REVIEW

File: scripts/staging/tSNE.dml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+    P = x2p(X, 1.0e-5, 20.0)
+    P = P*4
+    Y = rand(rows=n, cols=d, pdf="normal")
+    C = matrix(0, rows=max_iter, cols=1)
+
+    for (itr in 1:max_iter) {
+      D = distance_matrix(Y)
+      Z = 1/(D + 1)
+      parfor (j in 1:n) {
```
Comment (niketanpansare) [inline scripts/staging/tSNE.dml]: Thanks Imran. Minor suggestion when you run performance experiments: You can vectorize the parfor which might help improve the performance.
``` python
Z = Z * (diag(matrix(-1, rows=nrow(Z), cols=1)) + 1)
```

**Comment by niketanpansare** [comment]
Thanks Imran. This is awesome !! May be later, we can create a ipython notebook that guides a new user using "Merck Visualization Challenge": https://www.kaggle.com/c/MerckActivity/prospector#186 :+1:

File: scripts/staging/tSNE.dml (hunk 1 above)
Comment (iyounus) [inline scripts/staging/tSNE.dml]: Thanks. I'll make changes as you suggested.

Comment (mboehm7) [inline scripts/staging/tSNE.dml]: that's a very nice vectorization; just one thing to keep in mind: if Z contains NaN or +-INF the results would not be equivalent. However, as we have limited NaN support (other than primitives to removed them), it should be ok.

**Comment by deroneriksson** [comment]
@niketanpansare @mboehm7 Is this ready to be merged into staging directory?

**Comment by iyounus** [comment]
@deroneriksson I'm updating the gradient descent algorithm. I'll update the code today or tomorrow.

**Comment by deroneriksson** [comment]
Thank you for the update @iyounus!

File: scripts/staging/tSNE.dml [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    P = exp(-D * beta)
+    P = P * ZERODIAG
+    sum_Pi = rowSums(P)
+    W = rowSums(P * D)
+    Ws = W/sum_Pi
+    H = log(sum_Pi) + beta * Ws
+    P = P/sum_Pi
+    Hdiff = H - logU
+
+    for(i in 1:n) {
```
Comment (niketanpansare) [inline scripts/staging/tSNE.dml]: @iyounus I tried to matricize this for loop using following DML:
```python
betamin = betamin + (Hdiff > 0)*beta
betamax = betamax + (Hdiff <= 0)*beta
beta = 2*(Hdiff > 0)*(betamax == INF)*beta + (Hdiff > 0)*(betamax != INF)*(beta - betamax)/2 + (Hdiff <= 0)*(betamin == 0)*beta/2 + (Hdiff <= 0)*(betamin != 0)*(beta + betamin)/2
```
I wrote a sample DML to test the performance and also accuracy:
```python
n = 100000
INF = 1.0e20
betamax = matrix(INF, rows=n, cols=1)
betamin = matrix(0, rows=n, cols=1)
beta = matrix(seq(1, n), rows=n, cols=1)
Hdiff = matrix(seq(1, n)-(n/2), rows=n, cols=1)
# INSERT for loop or above expression here
print(toString(beta[10:20,]))
print(sum(beta))
```

Based on initial comparison, the for loop version took 27 seconds and the matricized version took 0.226 seconds. 

Not saying that this is the bottleneck, but thought my suggestion could help :)

Comment (iyounus) [inline scripts/staging/tSNE.dml]: Thanks for the suggestion. I'll make these changes.

**Comment by deroneriksson** [comment]
@iyounus @dusenberrymw 
Can we go ahead and merge this to staging as a first step towards its eventual move to a top-level algorithm?

**Comment by iyounus** [comment]
I would like to get some input from other people on this. My understanding is that this may not scale at all or at least will perform much worse than the tree based implementations that are available. So, we may just drop this. BTW, due to some bugs in systemml, I was never able to run with larger data set. But those bugs have been fixed.

**Comment by dusenberrymw** [comment]
Can you try running this again with the latest master?  A lot has changed in the time since the last commit, and it would be good to see if this is still broken.

**Comment by deroneriksson** [comment]
@iyounus Have you had a chance to check performance and also see if it can run with larger data now?

**Comment by deroneriksson** [comment]
@iyounus @dusenberrymw @niketanpansare @nakul02 @prithvirajsen 
What's the status of this PR? Can this be moved to staging, or should it be closed? It is coming up on a year since the PR was originally created.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1600/

**Comment by iyounus** [comment]
I would recommend to close this PR. I strongly doubt this will scale well.

**Comment by deroneriksson** [comment]
Thank you @iyounus for looking into the scalability. Could you go ahead and close the PR?

**Comment by j143** [comment]
As @mboehm7 pointed out, "the high-level summary: the problematic expression is indeed `X %*% t(X)` (in function distance_matrix). Given the 60Kx784 input, this output matrix is 60Kx60K, i.e., 29GB in dense. "

Here, I would like to know whether we are using any *special property* of multiplication with self transpose such as symmetry or we are doing an *ordinary* matrix multiplication.

the only expensive functions here: 
1. `X %*% t(X)` 
2. `W %*% Y` 

if we are using *ordinary* matrix multiplication, we can try for an efficient way of implementing it. This will help resolve this [SYSTEMML-1437].

cc @niketanpansare

**Comment by j143** [comment]
ping @niketanpansare . Please comment on above. Thanks. :smile:

**Comment by niketanpansare** [comment]
@j143 If possible, can you please provide statistics with a sample run along with steps to reproduce it at my end  ? This would help me comment on performance-related questions. Please set the `systemml.stats.finegrained` flag :)

**Comment by j143** [comment]
@niketanpansare Imran Younus has tested this script( only the first `x2p` function) with following command on spark, with 60k points of MNIST data.
```
spark-submit --master=spark://rr-ram4.softlayer.com:7077 --conf spark.executor.memory=120g --conf spark.driver.memory=80g /home/iyounus/git/incubator-systemml/target/SystemML.jar -f /home/iyounus/git/incubator-systemml/scripts/staging/tSNE.dml -exec hybrid_spark -nvargs INPUT=data/mnist_train_no_labels.csv OUT=data/P.csv
```
[log file](https://issues.apache.org/jira/secure/attachment/12830438/out_2016_09_26_10.log)

**Comment by niketanpansare** [comment]
Please do the following on your cluster:
1. Increase the ulimit on **all nodes**: `ulimit -n 65000`
2. `wget https://raw.githubusercontent.com/apache/systemml/master/conf/SystemML-config.xml.template`, `mv SystemML-config.xml.template SystemML-config.xml` and set `systemml.stats.finegrained` in that file to `true`.
3. Rerun your command and send statistics :)

**Pull request closed without merging**